### PR TITLE
wip: ceph: update helm chart for Ceph clusters in different namespaces

### PR DESCRIPTION
**Pull request by binoue**
This PR lets Helm users create necessary resources to have
Ceph clusters in different namespaces.

Signed-off-by: binoue <banji-inoue@cybozu.co.jp>

**Description of your changes:**
Update `values.yaml` to be able to specify namespaces for Ceph clusters and create necessary resources in each namespaces.

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test ceph]
[test full]

### CODE REVIEW

**Comment by BlaineEXE** [comment]
Do you have thoughts on this PR, @binoue ?

**Comment by binoue** [comment]
@BlaineEXE @travisn 

I also extended helm tests for multi clusters. Could you review this? 

Thanks!

File: cluster/charts/rook-ceph/templates/serviceaccount.yaml [hunk 1]
```diff
@@ -8,71 +13,83 @@ metadata:
     operator: rook
     storage-backend: ceph
     chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
-{{ template "imagePullSecrets" . }}
+{{ template "imagePullSecrets" $ }}
```
Comment (travisn) [inline cluster/charts/rook-ceph/templates/serviceaccount.yaml:16]: What is the difference between `$` and `.` with the template?

File: build/makelib/helm.mk [hunk 2]
```diff
@@ -43,6 +43,9 @@ $(HELM_OUTPUT_DIR)/$(1)-$(VERSION).tgz: $(HELM) $(HELM_OUTPUT_DIR) $(shell find
 	@echo === helm package $(1)
 	@cp -r $(HELM_CHARTS_DIR)/$(1) $(OUTPUT_DIR)
 	@$(SED_CMD) 's|VERSION|$(VERSION)|g' $(OUTPUT_DIR)/$(1)/values.yaml
+	@$(SED_CMD) 's|# clusterNamespaces:|clusterNamesspaces:|g' $(OUTPUT_DIR)/$(1)/values.yaml
```
Comment (travisn) [inline build/makelib/helm.mk:46]: What are the makefile changes for? Local testing? We wouldn't want to enable cluster-ns1 and cluster-ns2 in the build.

File: cluster/charts/rook-ceph/values.yaml [hunk 3]
```diff
@@ -321,3 +321,8 @@ discoverDaemonUdev:
 
 # Whether the OBC provisioner should watch on the operator namespace or not, if not the namespace of the cluster will be used
 enableOBCWatchOperatorNamespace: true
+
+# clusterNamespaces option allowes to create necessary resources in order to create Ceph clusters in the specified namespaces.
+# clusterNamesspaces:
+# - cluster-ns1
```
Comment (travisn) [inline cluster/charts/rook-ceph/values.yaml:327]: For the example, how about add a single example for `rook-ceph`?

File: cluster/charts/rook-ceph/values.yaml [hunk 4]
```diff
@@ -321,3 +321,8 @@ discoverDaemonUdev:
 
 # Whether the OBC provisioner should watch on the operator namespace or not, if not the namespace of the cluster will be used
 enableOBCWatchOperatorNamespace: true
+
+# clusterNamespaces option allowes to create necessary resources in order to create Ceph clusters in the specified namespaces.
+# clusterNamesspaces:
```
Comment (travisn) [inline cluster/charts/rook-ceph/values.yaml:326]:
```suggestion
# clusterNamespaces:
```

File: tests/framework/installer/ceph_installer.go [hunk 5]
```diff
@@ -396,20 +396,20 @@ func (h *CephInstaller) InstallRook(namespace, storeType string, usePVC bool, st
...
-		err = h.CreateRookOperatorViaHelm(namespace, settings)
+		err = h.CreateRookOperatorViaHelm(onamespace, settings)
 		if err != nil {
 			logger.Errorf("Rook Operator not installed ,error -> %v", err)
 			return false, err
 
 		}
 	} else {
-		onamespace = SystemNamespace(namespace)
+		onamespace = SystemNamespace(onamespace)
```
Comment (travisn) [inline tests/framework/installer/ceph_installer.go:412]: It would be better if `SystemNamespace()` were called when the test is first initialized. Then we could cleanup the call to `SystemNamespace()` in other places throughout the tests. I'll add an example below.

File: tests/integration/ceph_upgrade_test.go [hunk 6]
```diff
@@ -74,7 +74,8 @@ func (s *UpgradeSuite) SetupSuite() {
 	s.namespace = "upgrade-ns"
 	upgradeTestCluster := TestCluster{
 		clusterName:             s.namespace,
-		namespace:               s.namespace,
+		operatorNamespace:       s.namespace,
+		clusterNamespaces:       []string{s.namespace},
```
Comment (travisn) [inline tests/integration/ceph_upgrade_test.go:78]: Here is the example I mentioned above:
```
	upgradeTestCluster := TestCluster{
		clusterName:             s.namespace,
		namespace:               s.namespace,
		operatorNamespace:       installer.SystemNamespace(s.namespace),
		clusterNamespaces:       []string{s.namespace},
```

File: tests/framework/installer/ceph_installer.go [hunk 7]
```diff
@@ -396,20 +396,20 @@ func (h *CephInstaller) InstallRook(namespace, storeType string, usePVC bool, st
...
 	startDiscovery := true
-	onamespace := namespace
+	onamespace := operatorNamespace
 	// Create rook operator
 	if h.useHelm {
 		// disable the discovery daemonset with the helm chart
 		settings := "enableDiscoveryDaemon=false"
 		startDiscovery = false
-		err = h.CreateRookOperatorViaHelm(namespace, settings)
+		err = h.CreateRookOperatorViaHelm(onamespace, settings)
```
Comment (travisn) [inline tests/framework/installer/ceph_installer.go:405]: The helm chart is the only one that needs added support for multiple clusters, right? Looks like the `for` loop on line 428 should be here instead.

File: tests/framework/installer/ceph_installer.go [hunk 8]
```diff
@@ -425,12 +425,14 @@ func (h *CephInstaller) InstallRook(namespace, storeType string, usePVC bool, st
 	}
 
 	// Create rook cluster
-	err = h.CreateRookCluster(namespace, onamespace, storeType, usePVC, storageClassName,
-		cephv1.MonSpec{Count: mon.Count, AllowMultiplePerNode: mon.AllowMultiplePerNode}, startWithAllNodes,
-		skipOSDCreation, h.CephVersion)
-	if err != nil {
-		logger.Errorf("Rook cluster %s not installed, error -> %v", namespace, err)
-		return false, err
+	for _, namespace := range clusterNamespaces {
```
Comment (travisn) [inline tests/framework/installer/ceph_installer.go:428]: The tests don't need to support multiple namespaces for the non-helm case, except for the multi-cluster test suite, which is already implemented separately.

File: cluster/charts/rook-ceph/templates/clusterrolebinding.yaml [hunk 9]
```diff
@@ -18,6 +23,7 @@ subjects:
   namespace: {{ .Release.Namespace }}
 ---
 # Allow the ceph mgr to access cluster-wide resources necessary for the mgr modules
+{{- range  $_, $n := $clusterNamespaces }}
```
[the remark this answers is not in the document]
Comment (binoue) [reply cluster/charts/rook-ceph/templates/clusterrolebinding.yaml:26]: Sure! I will fix them.

File: cluster/charts/rook-ceph/templates/serviceaccount.yaml (hunk 1 above)
Comment (binoue) [reply cluster/charts/rook-ceph/templates/serviceaccount.yaml:16]: The change is for pointing root context.

As shown in [the official document](https://v2.helm.sh/docs/chart_template_guide/#variables), inside of looping, the current context(i.e. "." ) is overwritten and .Values.{something} can not be used.
> However, there is one variable that is always global - $ - this variable will always point to the root context. This can be very useful when you are looping in a range need to know the chart's release name.

Therefore, inside of the loop, I used the root context, "$", because, as it is root,  $.Values.{something} is always valid.

File: build/makelib/helm.mk (hunk 2 above)
Comment (binoue) [reply build/makelib/helm.mk]: >  Local testing?

Yes, the change was for local testing

> We wouldn't want to enable cluster-ns1 and cluster-ns2 in the build.

Okay I understood. Then, I will modify helm tests for passing values for `--set` option which would override values when installing chart.

`--set` option: https://v2.helm.sh/docs/helm/#helm-install

**Comment by binoue** [comment]
@travisn 

> * I would still consider recommending that an operator be run in the namespace with each cluster so that the operator can respond more quickly to reconcile requests for clusters. With a single operator, only a single cluster can reconcile at a time. It might just introduce delays for updates to the cluster CR on larger clusters. But for small clusters the delay would be small.

In order that the operator can respond more quickly, creating one operator for one ceph cluster is a possible solution.

I think, for that purpose, setting the "MaxConcurrentReconciles" parameter for a controller in [this place](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/controller.go#L137) would work too.

What do you think?

**Comment by travisn** [comment]
> @travisn
> 
> > * I would still consider recommending that an operator be run in the namespace with each cluster so that the operator can respond more quickly to reconcile requests for clusters. With a single operator, only a single cluster can reconcile at a time. It might just introduce delays for updates to the cluster CR on larger clusters. But for small clusters the delay would be small.
> 
> In order that the operator can respond more quickly, creating one operator for one ceph cluster is a possible solution.

A significant benefit of the separate operator is really the ability to troubleshoot the cluster. Each cluster generates a lot of logging, so having the clusters mingled in the same operator log would be difficult to troubleshoot.

> 
> I think, for that purpose, setting the "MaxConcurrentReconciles" parameter for a controller in [this place](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/controller.go#L137) would work too.
> 
> What do you think?

The challenge with allowing multiple concurrent cluster reconciles is that it would introduce race conditions. To keep the operator simpler, we have always made the assumption that only a single cluster is handled at a time. If we introduced concurrency, we would have to completely review the code and look for race conditions. 

The concurrency could be more easily enabled for other controllers besides the cluster controller, but those controllers would need to be reviewed as well to confirm.

**Comment by binoue** [comment]
@travisn 

Thank you for your comment and sorry to reply very late.

> A significant benefit of the separate operator is really the ability to troubleshoot the cluster. Each cluster generates a lot of logging, so having the clusters mingled in the same operator log would be difficult to troubleshoot.

If rook will recommend that one rook operator for each Ceph cluster,  then this PR would be better to modify the helm chart to make it easier?

**Comment by travisn** [comment]
> @travisn
> 
> Thank you for your comment and sorry to reply very late.
> 
> > A significant benefit of the separate operator is really the ability to troubleshoot the cluster. Each cluster generates a lot of logging, so having the clusters mingled in the same operator log would be difficult to troubleshoot.
> 
> If rook will recommend that a rook operator for each Ceph cluster, then this PR would be better to update the helm chart to be able to create multiple sets of a rook operator and a ceph cluster?
> 
> If so, I will modify the content of this PR.

It is still a valid scenario to have a single operator for multiple clusters. We just won't configure them concurrently as mentioned above. This PR is still valuable to enable supporting multiple clusters for a single operator.

The next step after this PR would be to confirm how to install a separate helm chart in case a separate operator is desired.

**Comment by binoue** [comment]
@travisn 

It takes long time, but I fixed this PR and finally confirmed that this passes Rook CI in my local environment.
However, I used two devices for two ceph cluster [here](https://github.com/rook/rook/blob/c350999af6be4a13dc12d0e9ab40c07da740635c/tests/scripts/localPathPV.sh#L4-L10) in this PR, so the Rook CI instance needs to have one more device.

Would you mind adding another device for the CI instance?

**Comment by travisn** [comment]
> @travisn
> 
> It takes long time, but I fixed this PR and finally confirmed that this passes Rook CI in my local environment.
> However, I used two devices for two ceph cluster [here](https://github.com/rook/rook/blob/2b80f0fed80292a523c06003ebf1bf1f0e75d5c9/tests/scripts/localPathPV.sh#L4-L10) in this PR, so the Rook CI instance needs to have one more device.
> 
> Would you mind adding another device for the CI instance?

@binoue Great to hear it's working locally! Yes, we can add another device to the CI, it might just take a few days to coordinate. I'll get back to you soon on that.

**Comment by binoue** [comment]
@travisn 

> Yes, we can add another device to the CI, it might just take a few days to coordinate. I'll get back to you soon on that.

That is amazing! Thank you so much!

File: cluster/charts/rook-ceph/values.yaml (hunk 3 above)
Comment (BlaineEXE) [reply cluster/charts/rook-ceph/values.yaml:326]: I think that with these changes, we should have `rook-ceph` be the default set option, which will instruct helm to create the resources that will allow the cluster to run in the same namespace as the operator. If a user were to remove `rook-ceph` from the options here, attempts to create a custom Ceph resources in the same namespace as the operator should fail. Does this make sense @travis?

Comment (travisn) [reply cluster/charts/rook-ceph/values.yaml:326]: Whether the operator is in the same namespace as a cluster doesn't seem like it would matter. The operator will run in whatever namespace is specified with the `helm install --namespace`, and the clusters can be created in the namespaces according to this list. If there are no namespaces in this list, then creating a cluster would always fail.

Comment (BlaineEXE) [reply cluster/charts/rook-ceph/values.yaml]: I think we're on the same page here. It sounds like you're describing what I was suggesting is my expectation. It sounds like that is what happens with this PR. :+1:

**Comment by binoue** [comment]
@travisn 

> Yes, we can add another device to the CI, it might just take a few days to coordinate. I'll get back to you soon on that.

Hi, is there any news on my request?:D

**Comment by travisn** [comment]
> @travisn
> 
> > Yes, we can add another device to the CI, it might just take a few days to coordinate. I'll get back to you soon on that.
> 
> Hi, is there any news on my request?:D

Hi, we have the second disk already. There was a slack discussion about it, but I missed responding to this PR. For example, in the latest master [build log](https://jenkins.rook.io/blue/rest/organizations/jenkins/pipelines/rook/pipelines/rook/branches/master/runs/2072/nodes/67/steps/134/log/?start=0), I see that both `xvdb` and `xvdc` exist:
```
2020-07-15 04:46:24.467541 I | testutil: Running tests/scripts/localPathPV.sh [/dev/xvdc]
NAME    MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
loop0     7:0    0 89.1M  1 loop /snap/core/8268
loop1     7:1    0   18M  1 loop /snap/amazon-ssm-agent/1480
loop2     7:2    0 28.1M  1 loop /snap/amazon-ssm-agent/2012
loop3     7:3    0 96.5M  1 loop /snap/core/9436
xvda    202:0    0  100G  0 disk 
└─xvda1 202:1    0  100G  0 part /
xvdb    202:16   0   10G  0 disk 
xvdc    202:32   0   10G  0 disk 
```
Let me know if that works for you.

**Comment by binoue** [comment]
@travisn 

Amazing! Thank you for answering!:D

**Comment by travisn** [comment]
@binoue Can you summarize the latest with this PR? Are you still investigating getting the integration tests to pass? The integration tests can be a challenge, so do let me know if you would like any input. Apologies for the delay in attention for this PR.

**Comment by llamerada-jp** [comment]
@travisn 
Thank you for your follow. I took over this PR from binoue.

I have waited to stabilize CI. I rebased this PR the latest master branch and found that some minor modifications are needed. I will request a review again after fixing these problems.

**Comment by Jasper-Ben** [comment]
activity!

**Comment by travisn** [comment]
@binoue Are you still pursuing this PR? Since #6809, installing to alternate or multiple namespaces is more easily done, at least with the raw manifests instead of the helm chart. Currently the helm chart can install to an alternate namespace, just not multiple namespaces. Is this really a requirement for users?

**Comment by binoue** [comment]
@travisn 

> @binoue Are you still pursuing this PR? Since #6809, installing to alternate or multiple namespaces is more easily done, at least with the raw manifests instead of the helm chart. Currently the helm chart can install to an alternate namespace, just not multiple namespaces. Is this really a requirement for users?

Yes as you have said, I think this PR is not much needed.
So, I will close this PR, but thank you for a lot of comments!
